src/interfaces/csp_if_eth_pbuf.c

```diff
@@ -6,118 +6,104 @@
+#define PBUF_TIMEOUT_MS 1000
 
-csp_packet_t * csp_if_eth_pbuf_find(csp_packet_t ** plist, uint32_t pbuf_id) {
+void csp_eth_pbuf_free(csp_eth_interface_data_t * ifdata, csp_packet_t * buffer, int buf_free, int * task_woken) {
 
-    csp_packet_t * packet = *plist;
-    while(packet) {
-        if (packet->cfpid == pbuf_id) {
-            return packet;
-        }
-        packet = packet->next;
-    }    
-    return packet;
+	csp_packet_t * packet = ifdata->pbufs;
+	csp_packet_t * prev = NULL;
 
-}
-
-void csp_if_eth_pbuf_insert(csp_packet_t ** plist, csp_packet_t * packet) {
+	while (packet) {
 
-    if (*plist) {
-        packet->next = *plist;
-    }
-    *plist = packet;
+		/* Perform cleanup in used pbufs */
+		if (packet == buffer) {
 
-}
+			/* Erase from list prev->next = next */
+			if (prev) {
+				prev->next = packet->next;
+			} else {
+				ifdata->pbufs = packet->next;
+			}
 
-csp_packet_t * csp_if_eth_pbuf_get(csp_packet_t ** plist, uint32_t pbuf_id, int * task_woken) {
+			if (buf_free) {
+				if (task_woken == NULL) {
+					csp_buffer_free(packet);
+				} else {
+					csp_buffer_free_isr(packet);
+				}
+			}
+		}
 
-    csp_packet_t * packet = csp_if_eth_pbuf_find(plist, pbuf_id);
+		prev = packet;
+		packet = packet->next;
+	}
 
-    if (packet) {
-    	packet->last_used = csp_get_ms();
-        return packet;
-    }
-
-    if (!packet) {
-        packet = (task_woken) ? csp_buffer_get_isr(0) : csp_buffer_get(0);
-    }
+}
 
-    if (!packet) {
-        /* No free packet */
-        return NULL;
-    }
+csp_packet_t * csp_eth_pbuf_new(csp_eth_interface_data_t * ifdata, uint32_t id, uint32_t now, int * task_woken) {
 
-	csp_id_setup_rx(packet);
+	csp_eth_pbuf_cleanup(ifdata, now, task_woken);
 
-    /* Existing cfpid and rx_count fields are used */ 
-    packet->cfpid = pbuf_id;
-    packet->rx_count = 0;
-	packet->last_used = (task_woken) ? csp_get_ms_isr() : csp_get_ms();
+	csp_packet_t * packet = (task_woken) ? csp_buffer_get_isr(0) : csp_buffer_get(0);
+	if (packet == NULL) {
+		return NULL;
+	}
 
-    packet->next = 0;
-    csp_if_eth_pbuf_insert(plist, packet);
+	packet->last_used = now;
+	packet->cfpid = id;
+	packet->frame_length = 0;
 
-    return packet;
+	/* Insert at beginning, because easy */
+	packet->next = ifdata->pbufs;
+	ifdata->pbufs = packet;
 
+	return packet;
 }
 
-void csp_if_eth_pbuf_remove(csp_packet_t ** plist, csp_packet_t * packet) {
+void csp_eth_pbuf_cleanup(csp_eth_interface_data_t * ifdata, uint32_t now, int * task_woken) {
 
-    csp_packet_t * prev = 0;
-    csp_packet_t * p = *plist;
-    while(p && (p != packet)) {
-        prev = p;
-        p = p->next;
-    }   
+	csp_packet_t * packet = ifdata->pbufs;
+	csp_packet_t * prev = NULL;
 
-    if (p) {
-        if (prev) {
-            prev->next = p->next;
-        } else {
-            *plist = p->next;
-        }
-    }
+	while (packet) {
 
-}
+		/* Perform cleanup in used pbufs */
+		if (now - packet->last_used > PBUF_TIMEOUT_MS) {
```

Comment:
I would add parentheses to this test, for clarity's sake, like this
`if ((now - packet->last_used) > PBUF_TIMEOUT_MS) {`

Reply:
I changed that a few places :)